### PR TITLE
arch/bl602：Add bl602 os adapter layer

### PR DESCRIPTION
**Pull request by SunJun8**
## Summary
Delete the os api call in the static link library and add the os adaptation layer

## Impact
None

## Testing
Compile passed
wapi scan passed

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by btashton — COMMENTED** [review]
@xiaoxiang781216 shouldn't the sleep functions here be using the kernel sig interfaces?

File: arch/risc-v/src/bl602/bl602_os_hal.c [hunk 1]
```diff
@@ -0,0 +1,1051 @@
...
+
+uint64_t bl_os_clock_gettime_ms(void)
+{
+  struct timespec ts;
+  clock_gettime(CLOCK_REALTIME, &ts);
+  return ts.tv_sec * 1000 + ts.tv_nsec / 1000000;
+}
+
+/****************************************************************************
+ * Name: esp_int_adpt_cb
```
Comment (btashton) [inline arch/risc-v/src/bl602/bl602_os_hal.c:734]: Wrong name

**Comment by SunJun8** [comment]
> @xiaoxiang781216 shouldn't the sleep functions here be using the kernel sig interfaces?

emm...right

File: arch/risc-v/src/bl602/bl602_os_hal.h [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+void bl_os_exit_critical(uint32_t level);
+
+#undef EXTERN
+#if defined(__cplusplus)
+}
+#endif
+
+#endif /* __ASSEMBLY__ */
+#endif /* __ARCH_RISCV_SRC_BL602_OS_HAL_H*/
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/bl602/bl602_os_hal.h:144]: add blank line

File: arch/risc-v/src/bl602/bl602_os_hal.h [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+ *
+ ****************************************************************************/
+
+#ifndef __ARCH_RISCV_SRC_BL602_OS_HAL_H
+#define __ARCH_RISCV_SRC_BL602_OS_HAL_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <assert.h>
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/bl602/bl602_os_hal.h:27]: bl602_os_hal.h  is shared by the binary library and the open source code, it's better to move all header file inclusion to .c except stdint.h

File: arch/risc-v/src/bl602/bl602_os_hal.c [hunk 4]
```diff
@@ -0,0 +1,1035 @@
...
+ *
+ * Returned Value:
+ *   True if success or false if fail
+ *
+ ****************************************************************************/
+
+int32_t bl_os_mutex_unlock(void *mutex_data)
+{
+  int ret;
+  pthread_mutex_t *mutex = (pthread_mutex_t *)mutex_data;
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/bl602/bl602_os_hal.c:882]: switch to semaphore

File: arch/risc-v/src/bl602/bl602_os_hal.c [hunk 5]
```diff
@@ -0,0 +1,1035 @@
...
+ *
+ * Input Parameters:
+ *
+ * Returned Value:
+ *
+ ****************************************************************************/
+
+void *bl_os_timer_create(void *func, void *argv)
+{
+  void *timerid;
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/bl602/bl602_os_hal.c:603]: kernel should use wdog or work with delay

File: arch/risc-v/src/bl602/bl602_os_hal.c (hunk 4 above)
Comment (SunJun8) [reply arch/risc-v/src/bl602/bl602_os_hal.c:882]: Why is it not recommended to use pthread？

Comment (xiaoxiang781216) [reply arch/risc-v/src/bl602/bl602_os_hal.c:882]: pthread API is only for userspace, all kernel component should use semaphore.

File: arch/risc-v/src/bl602/bl602_os_hal.c (hunk 5 above)
Comment (SunJun8) [reply arch/risc-v/src/bl602/bl602_os_hal.c:603]: Is timer api only used in userspace?or for other reasons

Comment (xiaoxiang781216) [reply arch/risc-v/src/bl602/bl602_os_hal.c:603]: It isn't safe to use timer api in the kernel space.

Comment (xiaoxiang781216) [reply arch/risc-v/src/bl602/bl602_os_hal.c:603]: If you call the API designed for userspace from kernel component, you may hit some mysterious bugs.

Comment (SunJun8) [reply arch/risc-v/src/bl602/bl602_os_hal.c:603]: > If you call the API designed for userspace from kernel component, you may hit some mysterious bugs.

Got it, thanks

File: arch/risc-v/src/bl602/bl602_os_hal.c [hunk 6]
```diff
@@ -0,0 +1,1259 @@
...
+
+  tmp = sizeof(sem_t);
+  sem = (sem_t *)kmm_malloc(tmp);
+  if (!sem)
+    {
+      wlerr("ERROR: Failed to alloc %d memory\n", tmp);
+      return NULL;
+    }
+
+  ret = nxsem_init(sem, 0, 0);
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/bl602/bl602_os_hal.c:1028]: nxsem_init(sem, 0, 1)

File: arch/risc-v/src/bl602/bl602_os_hal.c [hunk 7]
```diff
@@ -0,0 +1,1259 @@
...
+ *   line  - assert line
+ *   func  - assert function
+ *   expr  - assert condition
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+void __assert_func(const char *file, int line,
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/bl602/bl602_os_hal.c:156]: bl_assert_func

File: arch/risc-v/src/bl602/bl602_os_hal.c [hunk 8]
```diff
@@ -0,0 +1,1259 @@
...
+  ._zalloc = bl_os_zalloc,
+  ._get_time_ms = bl_os_clock_gettime_ms
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: __assert_func
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/bl602/bl602_os_hal.c:140]: bl_assert_func

File: arch/risc-v/src/bl602/bl602_os_hal.h [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+#define __ARCH_RISCV_SRC_BL602_OS_HAL_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <stdint.h>
+
+#ifdef CONFIG_BL602_WIRELESS
+#include "wifi_manager/bl_os_adapter.h"
+#endif
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/bl602/bl602_os_hal.h:31]: why include this file? I don't see there are any special type which come from bl_os_adapter.h

Comment (SunJun8) [reply arch/risc-v/src/bl602/bl602_os_hal.h:31]: obtain string`BL_OS_ADAPTER_VERSION`

Comment (xiaoxiang781216) [reply arch/risc-v/src/bl602/bl602_os_hal.h:31]: But it's used in bl60_os_hal.c, so the inclusion should move to that file.

File: arch/risc-v/src/bl602/bl602_os_hal.h [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+#define EXTERN extern
+#endif
+
+/****************************************************************************
+ * Definition
+ ****************************************************************************/
+
+#define OS_HPWORK HPWORK
+#define OS_LPWORK LPWORK
+
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/bl602/bl602_os_hal.h:54]: remove too

File: arch/risc-v/src/bl602/bl602_os_hal.h [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+extern void *__attribute__((weak)) _wifi_log_flag;
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/bl602/bl602_os_hal.h:55]: move to c file

**Review by xiaoxiang781216 — COMMENTED** [review]
@SunJun8 please reorg your patch into the reasonable series.

**Comment by SunJun8** [comment]
> @SunJun8 please reorg your patch into the reasonable series.

rebase complete

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/risc-v/src/bl602/bl602_netdev.c [hunk 12]
```diff
@@ -2120,10 +2122,12 @@ void bl602_net_event(int evt, int val)
     case CODE_WIFI_ON_DISCONNECT:
       do
         {
-          struct bl602_net_driver_s *priv = &g_bl602_net[0];
+          /* struct bl602_net_driver_s *priv = &g_bl602_net[0]; */
+
           if (g_state.sta_connected == 1)
             {
-              netdev_carrier_off(&priv->net_dev);
+              /* netdev_carrier_off(&priv->net_dev); */
```
Comment (normanr) [inline arch/risc-v/src/bl602/bl602_netdev.c:2129]: @SunJun8: Why this change? It seems to have been introduced in https://github.com/apache/incubator-nuttx/commit/79d35067fb9b03f52e0d8e383748a2c306ebf451, but I don't understand why.

Comment (SunJun8) [reply arch/risc-v/src/bl602/bl602_netdev.c:2129]: This looks like a change introduced due to a requirement. There may be some issues that need to be resolved at the time.

This change is outdated.

File: arch/risc-v/src/bl602/bl602_os_hal.h [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+
+void bl_os_log_write(uint32_t level,
+                     const char *tag,
+                     const char *file,
+                     int line,
+                     const char *format,
+                     ...);
+
+  #undef EXTERN
+  #if defined(__cplusplus)
```
Comment (pkarashchenko) [inline arch/risc-v/src/bl602/bl602_os_hal.h:181]: remove 2 spaces on this section

File: arch/risc-v/src/bl602/bl602_os_hal.h [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+}
+#endif
+
+#endif /* __ASSEMBLY__ */
+#endif /* __ARCH_RISCV_SRC_BL602_OS_HAL_H*/
```
Comment (pkarashchenko) [inline arch/risc-v/src/bl602/bl602_os_hal.h:186]: `HAL_H */`

File: arch/risc-v/src/bl602/bl602_os_hal.c [hunk 15]
```diff
@@ -0,0 +1,1606 @@
...
+      {
+        vsyslog(LOG_WARNING, format, args);
+        break;
+      }
+
+    case LOG_LEVEL_INFO:
+      {
+        vsyslog(LOG_INFO, format, args);
+        break;
+      }
```
Comment (pkarashchenko) [inline arch/risc-v/src/bl602/bl602_os_hal.c:1569]: add 2 spaces for this block and add `default` case handling

File: arch/risc-v/src/bl602/bl602_os_hal.c [hunk 16]
```diff
@@ -0,0 +1,1606 @@
...
+{
+  if (!work)
+    {
+      return -EINVAL;
+    }
+
+  if (!work_available((struct work_s *)work))
+    {
+      return -EINVAL;
+    }
```
Comment (pkarashchenko) [inline arch/risc-v/src/bl602/bl602_os_hal.c:1131]: Optional: can be merged to one if condition

File: arch/risc-v/src/bl602/bl602_os_hal.c [hunk 17]
```diff
@@ -0,0 +1,1606 @@
...
```
Comment (pkarashchenko) [inline arch/risc-v/src/bl602/bl602_os_hal.c:1102]: Optional: can be merged to one if condition

File: arch/risc-v/src/bl602/bl602_os_hal.c [hunk 18]
```diff
@@ -0,0 +1,1606 @@
...
+                                 mq_adpt->msgsize,
+                                 &prio,
+                                 &timeout);
+      if (ret < 0)
+        {
+          wlerr("ERROR: Failed to timedreceive from mqueue error=%d\n", ret);
+        }
+    }
+
+  return ret < 0 ? false : true;
```
Comment (pkarashchenko) [inline arch/risc-v/src/bl602/bl602_os_hal.c:885]: `return !(ret < 0);`

File: arch/risc-v/src/bl602/bl602_netdev.c (hunk 12 above)
Comment (normanr) [reply arch/risc-v/src/bl602/bl602_netdev.c:2129]: It was included in https://github.com/apache/incubator-nuttx/commit/d489392d08e550e828146395d4e945de9952e6e3 (which was merged) though. To me it makes more sense to _have_ the call to `netdev_carrier_off` in `CODE_WIFI_ON_DISCONNECT` (so that userspace can detect the disconnection), and to _not_ have the call to `wifi_mgmr_sta_autoconnect_disable` in `CODE_WIFI_ON_CONNECTED` (so that `CODE_WIFI_CMD_RECONNECT` can trigger correctly). Thoughts?

Comment (SunJun8) [reply arch/risc-v/src/bl602/bl602_netdev.c:2129]: yes... `netdev_carrier_off`should be called here and it is used like this in unmerged internal code.

Comment (SunJun8) [reply arch/risc-v/src/bl602/bl602_netdev.c:2129]: I will update this part of the code in due course

**Comment by SunJun8** [comment]
@pkarashchenko thanks for your review, and I will update this part of the code in due course
